♻️ refactor : AI 적성 검사 로직 및 구조 개선 (횟수 제한, 캐시 활용, JSON 파싱, 신뢰도 적용)

sooyeoneo opened the pull request:
## ☀️ 작업 목적

* 온보딩과 마이페이지 적성 검사를 구분하여 정확한 횟수 제한을 적용
* AI 응답 파싱을 JSON 구조화하여 신뢰도 기반 점수 계산이 가능하도록 개선
* Redis 캐시를 활용해 답변·점수·진행상태 관리를 강화하고, 불필요한 DB 파싱 제거

---

## 📄 주요 작업 내역
### 1. 적성 검사 횟수 제한 로직 개선 

* `UserAptitude` 엔티티 확장
  * `mypageTestCount` 필드: 마이페이지 검사 횟수 추적
  * `isOnboardingCompleted` 필드: 온보딩 완료 여부
* 온보딩/마이페이지 검사 구분
  * 온보딩 첫 검사는 횟수 제한 없음
  * 마이페이지 검사만 최대 5회 제한 (총 6회 가능)
* API 문서/에러 메시지 업데이트
  * `APTITUDE_TEST_LIMIT_EXCEEDED` 메시지 수정
  * 온보딩/마이페이지 정책 명시
* #55 #66

### 2. `AptitudeAiService` 및 캐시 구조 개선 

* `AptitudeCacheService` 기능 확장
  * `saveAnswer()`: 단계별 답변 저장
  * `updateAptitudeScore()`: 적성별 점수 누적
  * `getAllAptitudeScores()`: 전체 점수 조회
  * `deleteAllTestData()`: 데이터 정리
* `AptitudeAiService` 단순화
  * AI 호출 실패 시 fallback 키워드 분석
  * 로그 레벨 warn 적용
* `AptitudeService` 개선
  * Redis 점수 실시간 활용
  * DB 파싱 제거, 테스트 완료 후 데이터 자동 정리
* #55 #65 #67 #151

### 3. 테스트 환경 및 데이터 정리 로직 강화 

* HTTP 테스트 파일 개선
  * 온보딩/단계별/수동 선택 시나리오 분리
  * 동적 변수 경로 수정 및 콘솔 로그 추가
* Redis 데이터 정리 로직 강화
  * `deleteAllTestData()` 상세 로그 추가
  * 시작/완료 시점, 답변/점수 삭제 추적
  * 디버깅 로그 강화로 진행 상황 가시성 확보
 * #55 #65 #67 #151

### 4. AI 적성 분석 JSON 구조화 및 신뢰도 기반 점수 

* DTO 추가
  * `AiAptitudeAnalysisResponseDto`: 적성 타입, 신뢰도, 이유, 핵심요소
  * `AiQuestionGenerationResponseDto`: 동적 질문 생성 대비
* JSON 응답 기반 분석
  * 신뢰도 0.7 이상 → 2점, 미만 → 1점 가중치 적용
  * fallback 분석도 동일 객체 구조 유지
* 로그 추가: JSON 파싱/가중치 적용 확인
* #55 #64 #67 #151

### 5. 세션 타임아웃 설정 

* 적성 검사 세션 TTL을 30분으로 단축 → 테스트 집중도 및 리소스 최적화
* #55 #64

---

## 📆 수정 내역 (2025.09.03)
♻️ refactor : 코드 리뷰 반영 - AI 적성 분석 안정성 및 일관성 개선 

* `AiAptitudeAnalysisResponseDto`
  * `String[] keyFactors` → `List<String>` 변경
  * `Defensive Copy` 적용으로 null 안전성 향상
  * 다른 DTO들과 컬렉션 타입 통일
* `AptitudeService`
  * `toUpperCase()` 적용 → 소문자 입력 대응
  * 응답의 대소문자 구분 없이 안전 처리
* `AptitudeAiService`
  * `List.of()` 사용으로 fallback 로직 일관성 확보
* 효과: 프로젝트 일관성/안전성 향상, 유지보수성 강화

---

## 📷 테스트 결과 캡처
### JSON 파싱 및 신뢰도 기반 점수 누적 (Step 1~5)

#### Step 1 
<img width="1792" height="1137" alt="적성 AI 검사 JSON 파싱 구조화 로깅 테스트 2025-09-03 09 15 48" src="https://github.com/user-attachments/assets/f5d650ba-eb76-4e05-9023-d0374fd82311" />

#### Test log 1
<img width="1606" height="543" alt="적성 AI 검사 JSON 파싱 구조화 로깅 테스트 2025-09-03 09 15 20" src="https://github.com/user-attachments/assets/bb259b17-cc50-4de8-a770-4a5a7cbe9598" />

#### Step 2
<img width="1792" height="1137" alt="적성 AI 검사 JSON 파싱 구조화 로깅 테스트 2025-09-03 09 16 18" src="https://github.com/user-attachments/assets/adb59760-2826-4afb-8152-27c9d437c89c" />

#### Test log 2
<img width="1606" height="543" alt="적성 AI 검사 JSON 파싱 구조화 로깅 테스트 2025-09-03 09 16 10" src="https://github.com/user-attachments/assets/7f0736cb-f603-40c2-ba0c-10ba43177887" />

#### Step 3
<img width="1792" height="1137" alt="적성 AI 검사 JSON 파싱 구조화 로깅 테스트 2025-09-03 09 16 41" src="https://github.com/user-attachments/assets/90d7278f-cad9-4cc7-be52-ae0f2b87a50d" />

#### Test log 3
<img width="1606" height="543" alt="적성 AI 검사 JSON 파싱 구조화 로깅 테스트 2025-09-03 09 16 35" src="https://github.com/user-attachments/assets/0f769601-7b05-4726-8800-7a6a979539a8" />

#### Step 4
<img width="1792" height="1137" alt="적성 AI 검사 JSON 파싱 구조화 로깅 테스트 2025-09-03 09 17 04" src="https://github.com/user-attachments/assets/522ea009-962c-4c3e-ad51-54cf1722f98c" />

#### Test log 4
<img width="1606" height="543" alt="적성 AI 검사 JSON 파싱 구조화 로깅 테스트 2025-09-03 09 16 57" src="https://github.com/user-attachments/assets/be0112b7-b1ba-46cc-a692-bb4a2e67d665" />

#### Step 5
<img width="1792" height="1137" alt="적성 AI 검사 JSON 파싱 구조화 로깅 테스트 2025-09-03 09 17 42" src="https://github.com/user-attachments/assets/312cc484-513a-4394-8604-5ab0e40b4542" />

#### Test log 5
<img width="1606" height="543" alt="적성 AI 검사 JSON 파싱 구조화 로깅 테스트 2025-09-03 09 17 35" src="https://github.com/user-attachments/assets/e003f1da-5e3a-4cb8-842f-858556ddf540" />

### OpenAI API 토큰 사용량 추적
<img width="1473" height="890" alt="JSON 파싱 구조화 테스트 후 토큰 사용량 2025-09-03 09 18 39" src="https://github.com/user-attachments/assets/05ab0340-8e82-454f-8e4b-44b7cacdc1cb" />

* 총 요청: 19회
* 입력 토큰: 6.1K
* 비용: $0.01 미만

---

## 🛠️ 추가 수정 작업 (TODO)

* `GlobalExceptionHandler` 통합: `CustomException` 처리 필요
* 프롬프트 고도화: JSON Schema 기반 응답 요청 및 validation 적용
* 프론트엔드 연동: 온보딩/마이페이지 검사 결과 UX 반영, 프로그램 예시 제공 API 확장
* 테스트 보강: JSON 파싱 실패/낮은 신뢰도/Redis TTL 만료 케이스 단위 테스트 추가

---

**리뷰 및 피드백 환영합니다! 🤗**

---

Note: The `Co-authored-by: @jiyoon0000` tag reflects code-review feedback,
not direct code implementation or PR document input.

sooyeoneo committed:
✨ feat : 적성 검사 Entity 및 Enum 구현

- UserAptitude Entity 구현
  - 회원별 적성 정보 저장 (1:1 관계)
  - 테스트 횟수 추적 (최대 5회 제한)
  - BaseEntity 상속으로 생성/수정 시간 자동 관리

- AptitudeTestHistory Entity 구현
  - 적성 검사 이력 저장 (N:1 관계)
  - 질문-답변-AI응답 기록

- AptitudeType Enum 구현
  - 5가지 적성 타입 정의 (자연친화, 역사문화, 예술창작, 커뮤니티, 디지털기술)
  - 각 타입별 제목과 설명 포함

#65

sooyeoneo committed:
✨ feat : 적성 검사 Repository 구현

- UserAptitudeRepository
  - 회원별 적성 조회 (findByMember)
  - 적성 존재 여부 확인 (existsByMember)

- AptitudeTestHistoryRepository
  - 회원별 검사 이력 조회 (오름차순 정렬)
  - 회원별 이력 전체 삭제

#65

sooyeoneo committed:
✨ feat : 적성 검사 DTO 구현

- RequestDto (요청)
  - AptitudeAnswerRequestDto: 사용자 답변 제출

- ResponseDto (응답)
  - AptitudeQuestionResponseDto: AI 질문 정보를 사용자에게 답변 응답
  - AptitudeTestStartResponseDto: 검사 시작 응답
  - AptitudeTextProgressResponseDto: 진행 상태 응답
  - AptitudeTestResultResponseDto: 최종 결과 응답
  - CanRetakeTestResponseDto: 재검사 가능 여부 응답

* 모든 ResponseDto는 불변 객체로 구현 (생성자 사용)

#64 #65

sooyeoneo committed:
✨ feat : 적성 검사 Service 구현

- AptitudeAiService
  - 5개 질문 Pool 정적 관리
  - AI 프롬프트 기반 답변 분석
  - 적성 타입별 점수 계산 및 최종 적성 결정
  - BaseAiService 상속으로 재시도 및 예외처리

- AptitudeService
  - 검사 시작/답변 제출/결과 저장 플로우
  - Redis 기반 검사 진행 상태 관리 (24시간 TTL)
  - 최대 5회 재검사 제한 로직
  - 수동 적성 선택 기능
  - CustomException 기반 예외처리

* @Transactional 메서드별 적용

#64 #65 #66

sooyeoneo committed:
✨ feat : 적성 검사 Controller 구현

- API 엔드포인트
  - POST /api/ai/aptitude/test/start: 검사 시작
  - POST /api/ai/aptitude/test/answer: 답변 제출
  - POST /api/ai/aptitude/select: 수동 적성 선택
  - GET /api/ai/aptitude/can-retake: 재검사 가능 여부

- Spring Security 적용
  - @PreAuthorize(\"hasRole('USER')\") 일반회원 권한 체크
  - MemberDetails 통한 회원 정보 획득

- CommonResponseDto 표준 응답 형식 적용

#63 #66

sooyeoneo committed:
🔧 add : Redis 범용 aiAptitudeRedisTemplate 빈 추가

- RedisTemplate<String, String> 빈 추가
- @Qualifier("aiAptitudeRedisTemplate") 빈에 추가 구분자를 붙여 명시적으로 지정
- 적성 검사 진행 상태 관리용
- 기존 blacklist, emailVerified 템플릿과 별도 관리

#64

sooyeoneo committed:
✅ test : 적성 검사 HTTP 테스트 및 AI 프롬프트 추가

- ai-aptitude.http
  - 적성 검사 시작/답변/선택/재검사 API 테스트
  - Authorization Bearer 토큰 기반

- aptitude-test.txt
  - 적성 분석 AI 프롬프트 추가

#67

sooyeoneo committed:
💡 comment : AiTestController 주석 처리 (추후 삭제 예정)

sooyeoneo committed:
💡 comment : 필요 없는 주석 삭제

sooyeoneo committed:
🔥 remove : @PreAuthorize("hasRole('USER')") 제거

sooyeoneo committed:
🔧 add : NOT_FOUND_TEST_PROGRESS, TOO_MANY_REQUESTS ErrorCode에 추가

sooyeoneo committed:
🔧 add : 적성 검사 AI 응답 분석에 정규식 패턴 매칭 로직 추가

- analyzeHistories() 메서드에 Pattern/Matcher 기반 분석 로직 적용
- 정규식으로 모든 적성 타입(name + title) OR 연결 패턴 생성
- AI 응답 내 복수 적성 키워드 추출 및 점수 계산 지원
- 향후 HashMap 최적화를 위한 TODO 주석 추가

Refs: #65

sooyeoneo committed:
Merge branch 'develop' into feature/ai-aptitude

# Conflicts:
#	src/main/java/com/salayo/locallifebackend/domain/ai/aptitude/service/AptitudeService.java
#	src/main/java/com/salayo/locallifebackend/global/config/RedisConfig.java
#	src/main/java/com/salayo/locallifebackend/global/error/ErrorCode.java

sooyeoneo committed:
Merge branch 'develop' into feature/ai-aptitude

sooyeoneo committed:
🔧 add : @PreAuthorize("hasRole('USER')") 추가

- AptitudeController에 @PreAuthorize("hasRole('USER')") 일반 회원 권한을 다시 추가

#55

sooyeoneo committed:
🔧 add : @RequireArgsConstructor 삭제하고, 생성자 직접 주입

- AptitudeController에 @RequireArgsConstructor 삭제하고, 생성자 직접 주입(Constructor injection)으로 의존성을 직접 전달

#55

sooyeoneo committed:
🔥 remove : 사용하지 않는 import 문 제거

sooyeoneo committed:
♻️ refactor : 적성 검사 캐시 로직 분리 및 CacheKeyPrefix로 통합 관리

- AptitudeCacheService 생성하여 Redis 캐시 로직 캡슐화
  - ReviewCacheService와 동일한 아키텍처 패턴 적용
  - 테스트 진행 상태 저장/조회/삭제 메서드 구현

- CacheKeyPrefix에 적성 검사 관련 모든 상수 통합
  - APTITUDE_MAX_TEST_COUNT: 최대 테스트 횟수 (5회)
  - APTITUDE_TOTAL_QUESTIONS: 전체 질문 개수 (5개)
  - APTITUDE_TEST: Redis 키 프리픽스 ("aptitude:test:")
  - APTITUDE_TEST_TTL_HOURS: 캐시 TTL (24시간)

- AptitudeService 리팩토링
  - 캐시 로직을 AptitudeCacheService로 위임
  - CacheKeyPrefix 상수 참조로 중앙 집중식 관리로 유지보수성 향상
  - AptitudeService와 AptitudeCacheService에서 CacheKeyPrefix 참조

#55 #65

sooyeoneo committed:
💡 comment : 주석 수정

sooyeoneo committed:
🐛 fix : 적성 검사 횟수 제한 에러코드 수정

- APTITUDE_TEST_LIMIT_EXCEEDED 에러코드 추가 (400 BAD_REQUEST)
- TOO_MANY_REQUESTS(429)는 Rate Limiting 전용으로 복원
- 비즈니스 로직 제약사항에 맞는 적절한 상태 코드 사용

#55 #65 #67

sooyeoneo committed:
📝 docs : 적성 검사 API Swagger 문서화 추가

- Controller 레벨
  - @Tag 추가: "Aptitude" 그룹화 및 설명
  - @Operation: 각 API 엔드포인트별 요약 및 상세 설명
  - @ApiResponses: HTTP 상태 코드별 응답 설명 (200, 201, 400, 401, 404)
  - @Parameter: 요청 파라미터 설명 및 예시값

- DTO 레벨
  - 모든 Request/Response DTO에 @Schema 추가
  - 필드별 설명, 예시값, 최소/최대값, 필수 여부 명시
  - AptitudeQuestionResponseDto: 질문 정보 및 예시 답변
  - AptitudeAnswerRequestDto: 답변 요청 필드 제약사항
  - CanRetakeTestResponseDto: 재검사 관련 정보

- Enum 레벨
  - AptitudeType에 @Schema 추가
  - 각 적성 타입별 상세 설명 포함

Swagger UI에서 적성 검사 API의 상세 문서 확인 가능

#55 #64 #65 #66

sooyeoneo committed:
🔧 add : @Valid 추가

sooyeoneo committed:
✏️ typos : 코드 줄 정리

sooyeoneo committed:
🔧 add : ResponseDto에 @Builder를 생성자 레벨에 추가

sooyeoneo committed:
♻️ refactor : 적성 검사 횟수 제한 로직 개선 - 온보딩/마이페이지 구분

- UserAptitude 엔티티 개선
  - mypageTestCount 필드 추가 (마이페이지 검사 횟수)
  - isOnboardingCompleted 필드 추가 (온보딩 완료 여부)
  - 온보딩/마이페이지별 업데이트 메서드 분리

- 적성 검사 제한 로직 변경
  - 온보딩 첫 검사는 횟수 제한 없음
  - 마이페이지에서만 최대 5회 제한 적용
  - 총 6회 검사 가능 (온보딩 1회 + 마이페이지 5회)

- API 문서 및 에러 메시지 업데이트
  - 온보딩과 마이페이지 검사 구분 명시
  - APTITUDE_TEST_LIMIT_EXCEEDED 메시지 수정

#55 #66

sooyeoneo committed:
Merge branch 'develop' into feature/ai-aptitude

# Conflicts:
#	src/main/java/com/salayo/locallifebackend/domain/ai/aptitude/controller/AptitudeController.java
#	src/main/java/com/salayo/locallifebackend/domain/ai/aptitude/dto/CanRetakeTestResponseDto.java
#	src/main/java/com/salayo/locallifebackend/domain/ai/aptitude/service/AptitudeService.java
#	src/main/java/com/salayo/locallifebackend/global/error/ErrorCode.java

sooyeoneo committed:
♻️ refactor : AptitudeAiService 구조 개선 및 캐시 활용 강화

- AptitudeCacheService에 답변 저장 및 적성 점수 관리 기능 추가
  - saveAnswer(): 단계별 답변 Redis 저장
  - updateAptitudeScore(): 적성별 점수 누적 관리
  - getAllAptitudeScores(): 전체 점수 조회
  - deleteAllTestData(): 테스트 데이터 정리

- AptitudeAiService 단순화 및 개선
  - AI 호출 실패 시 fallback 키워드 분석
  - 로그 레벨 log.warn 설정

- AptitudeService Redis 활용 개선
  - AI 분석 결과를 Redis에 실시간 저장
  - DB 텍스트 파싱 대신 Redis 점수 직접 활용
  - 테스트 완료 후 자동 데이터 정리

#55 #65 #67 #151

sooyeoneo committed:
♻️ refactor : AI 적성 검사 테스트 환경 및 데이터 정리 로직 개선

- HTTP 테스트 파일 개선
  - 동적 변수 처리를 위한 응답 구조 추가(aptitudeQuestionResponseDto 경로 수정)
  - AI 적성 검사 전체 플로우, 개별 단계별, 수동 선택 테스트 섹션 구분
  - 콘솔 로그 추가로 테스트 진행 상황 추적 가능

- Redis 데이터 정리 로직 강화
  - deleteAllTestData() 메소드에 상세 로그 추가
  - 각 데이터 타입별 삭제 결과 확인 (진행상태, 답변, 점수)
  - 테스트 시작 시 기존 데이터 정리로 중복 방지

- 디버깅 로그 추가
  - Redis 데이터 정리 시작/완료 로그
  - 답변 저장 및 점수 업데이트 과정 추적

#55 #65 #67 #151

sooyeoneo committed:
🏗 chore : AI 적성 검사 세션 타임아웃 30분 설정

#55 #64

sooyeoneo committed:
✨ feat : AI 적성 분석 JSON 파싱 구조화 및 신뢰도 기반 점수 적용

- AiAptitudeAnalysisResponseDto 생성 (적성 타입, 신뢰도, 이유, 핵심요소)
- AiQuestionGenerationResponseDto 생성 (추후 동적 질문 생성용)
- AptitudeAiService에서 구조화된 객체 반환 및 JSON 응답 프롬프트 개선
- 신뢰도 점수(0.7 기준)에 따른 가중치 적용 (높음: 2점, 낮음: 1점)
- Fallback 분석도 동일한 객체 구조로 통일
- 테스트용 로그 추가 (JSON 파싱, 가중치 적용 확인)

#55 #64 #67 #151

sooyeoneo committed:
Merge branch 'develop' into feature/ai-aptitude

j-hann reviewed (approved):
짱짱👍🤗

jiyoon0000 commented on src/main/java/com/salayo/locallifebackend/domain/ai/aptitude/service/AptitudeService.java:
```diff
@@ -217,4 +275,35 @@ private Map<AptitudeType, Integer> analyzeHistories(List<AptitudeTestHistory> hi
+	private void updateScoreFromAnalysis(Long memberId, AiAptitudeAnalysisResponseDto aiAnalysis) {
+		try {
+			// AI 분석 결과에서 적성 타입 추출
+			AptitudeType aptitudeType = AptitudeType.valueOf(aiAnalysis.getAptitudeType());
```
fallback이 되어있는 것 같긴한데 혹시 소문자가 들어올걸 대비해서 `toUpperCase()` 적용하는 건 어떨까요?🧐

`AptitudeType aptitudeType = AptitudeType.valueOf(aiAnalysis.getAptitudeType().toUpperCase());`

sooyeoneo replied:
바로 적용하였습니다! 감사합니다! 🥹

jiyoon0000 commented on src/main/java/com/salayo/locallifebackend/domain/ai/aptitude/dto/AiAptitudeAnalysisResponseDto.java:
```diff
@@ -0,0 +1,33 @@
+    private String reason; // 추천 이유
+    
+    @JsonProperty("key_factors")
+    private String[] keyFactors; // 핵심 판단 요소들
```
고정배열이라 String[]으로 둔걸까요??
아니라면 `private List<String> keyFactors;` 처럼 컬렉션을 사용하는 건 어떨까요?🤔

sooyeoneo replied:
좋은 의견 감사합니다!

처음에는 AI 응답 파싱 성능을 고려해서 `String[]`을 선택했습니다. 
배열 접근이 List 보다 미세하게 빠르고 AI 응답 구조가 고정적이라고 생각했습니다.
실제로는 아직 성능 차이를 고려할 만큼 트래픽이 많지 않기 때문에, `List<String>`이 더 적합한 것 같네요!

* 일관성: `AptitudeQuestionResponseDto`의 `exampleAnswers`도 List<String> 사용중
* 안전성: Fallback 처리와 null 체크 안전
* 확장성: GPT-4o mini가 때때로 1~5개의 가변 요소를 반환할 수 있음
* Spring: Jackson이 List를 더 유연하게 처리

프로젝트 전체 일관성과 유지보수성이 ns 단위 성능보다 중요하다는 것을 간과했네요.  
`List<String>`으로 수정하겠습니다! 감사합니다 ☺️

sooyeoneo committed:
♻️ refactor : 코드 리뷰 반영 - AI 적성 분석 안정성 및 일관성 개선

[변경 내용]
1. AiAptitudeAnalysisResponseDto 개선
   - String[] keyFactors → List<String> keyFactors 변경
   - 방어적 복사(Defensive Copy) 적용으로 null 안전성 향상
   - 프로젝트 내 다른 DTO들과 컬렉션 사용 통일

2. AptitudeService 안전성 강화
   - toUpperCase() 적용으로 소문자 입력 대응
   - AI 응답의 대소문자 구분 없이 안전하게 처리

3. AptitudeAiService fallback 로직 개선
   - List.of() 사용으로 일관성 유지
   - Fallback 시에도 동일한 자료구조 사용

[개선 효과]
- 프로젝트 전체 일관성 확보
- null 및 대소문자 안전성 향상
- Spring/Jackson 표준 준수
- 유지보수성 개선

Resolves: #158
Co-authored-by: jiyoon0000 <rlawldbs9128@gmail.com>
